# KIKA Materials Module - Complete Tutorial

This notebook demonstrates all the functionality of the KIKA materials module, which provides a general-purpose framework for defining, manipulating, and exporting material compositions for various Monte Carlo codes including MCNP and Serpent.

In [1]:
from kika.materials import Material, MaterialCollection, Nuclide, NuclideAccessor

## 1. Creating Materials

Materials can be created in several ways. The most basic is to create an empty material and add nuclides to it.

### 1.1 Basic Material Creation

In [2]:
# Create a material with ID=1
water = Material(id=1)

# Add hydrogen and oxygen (natural elements)
water.add_nuclide('H', 2.0, 'ao')  # 2 atoms of H
water.add_nuclide('O', 1.0, 'ao')  # 1 atom of O

# Set the name and density
water.set_name('Water')
water.set_density(1.0, 'g/cc')

print(water)

                 Material (ID: 1)                 

Name: Water
Density: 1 g/cc
Total Nuclides: 2

Fraction Type: Atomic

--------------------------------------------------
  ZAID   | Element  |    Fraction   
--------------------------------------------------
  1000   |    H     |  2.000000e+00 
  8000   |    O     |  1.000000e+00 
--------------------------------------------------

Available methods:
- .to_weight_fraction() - Convert material to weight fractions
- .to_atomic_fraction() - Convert material to atomic fractions
- .expand_natural_elements() - Expand natural elements to isotopes
- .to_integer_fractions() - Convert normalized fractions to integers
- .to_mcnp() - Export as MCNP material card
- .to_serpent() - Export as Serpent material card
- .copy(new_id) - Create a copy with a new material ID

Examples of accessing data:
- .nuclide['Fe56'] - Access nuclide by symbol
- .nuclide[26056] - Access nuclide by ZAID
- .nuclide['Fe'] or .nuclide[26000] - Access natural Fe


### 1.2 Adding Nuclides by ZAID

You can also add nuclides by their ZAID (ZZAAA format where ZZ is atomic number and AAA is mass number).

In [3]:
# Create enriched uranium
uranium = Material(id=2)
uranium.set_name('Enriched Uranium')

# Add by ZAID (92235 = U-235, 92238 = U-238)
uranium.add_nuclide(92235, 0.05, 'ao')  # 5% U-235
uranium.add_nuclide(92238, 0.95, 'ao')  # 95% U-238

uranium.set_density(19.1, 'g/cc')

print(uranium)

                 Material (ID: 2)                 

Name: Enriched Uranium
Density: 19.1 g/cc
Total Nuclides: 2

Fraction Type: Atomic

--------------------------------------------------
  ZAID   | Element  |    Fraction   
--------------------------------------------------
 92235   |    U     |  5.000000e-02 
 92238   |    U     |  9.500000e-01 
--------------------------------------------------

Available methods:
- .to_weight_fraction() - Convert material to weight fractions
- .to_atomic_fraction() - Convert material to atomic fractions
- .expand_natural_elements() - Expand natural elements to isotopes
- .to_integer_fractions() - Convert normalized fractions to integers
- .to_mcnp() - Export as MCNP material card
- .to_serpent() - Export as Serpent material card
- .copy(new_id) - Create a copy with a new material ID

Examples of accessing data:
- .nuclide['Fe56'] - Access nuclide by symbol
- .nuclide[26056] - Access nuclide by ZAID
- .nuclide['Fe'] or .nuclide[26000] - Access natura

### 1.3 Adding Natural Elements with add_element()

The `add_element()` method is a convenience function for adding natural elements.

In [4]:
# Create stainless steel (simplified)
steel = Material(id=3)
steel.set_name('Stainless Steel 304')

# Add natural elements
steel.add_element('Fe', 0.70, 'wo')  # 70% iron by weight
steel.add_element('Cr', 0.18, 'wo')  # 18% chromium
steel.add_element('Ni', 0.10, 'wo')  # 10% nickel
steel.add_element('C', 0.02, 'wo')   # 2% carbon

steel.set_density(8.0, 'g/cc')

print(steel)

                 Material (ID: 3)                 

Name: Stainless Steel 304
Density: 8 g/cc
Total Nuclides: 4

Fraction Type: Weight

--------------------------------------------------
  ZAID   | Element  |    Fraction   
--------------------------------------------------
  6000   |    C     |  2.000000e-02 
 24000   |    Cr    |  1.800000e-01 
 26000   |    Fe    |  7.000000e-01 
 28000   |    Ni    |  1.000000e-01 
--------------------------------------------------

Available methods:
- .to_weight_fraction() - Convert material to weight fractions
- .to_atomic_fraction() - Convert material to atomic fractions
- .expand_natural_elements() - Expand natural elements to isotopes
- .to_integer_fractions() - Convert normalized fractions to integers
- .to_mcnp() - Export as MCNP material card
- .to_serpent() - Export as Serpent material card
- .copy(new_id) - Create a copy with a new material ID

Examples of accessing data:
- .nuclide['Fe56'] - Access nuclide by symbol
- .nuclide[26056] - 

### 1.4 Setting Temperature

In [5]:
# Set temperature in Kelvin
water.set_temperature(300.0)  # 300 K (~27°C)
print(f"Water temperature: {water.temperature} K")

Water temperature: 300.0 K


## 2. Accessing Material Properties

### 2.1 Nuclide Access by Symbol or ZAID

In [6]:
# Access by symbol
h_nuclide = water.nuclide['H']
print(f"Hydrogen by symbol: ZAID={h_nuclide.zaid}, fraction={h_nuclide.fraction}")

# Access by ZAID
o_nuclide = water.nuclide[8000]  # 8000 = natural oxygen
print(f"Oxygen by ZAID: symbol={o_nuclide.symbol}, fraction={o_nuclide.fraction}")

# Check if nuclide exists
print(f"\nHas 'Fe'? {'Fe' in water.nuclide}")
print(f"Has 'H'? {'H' in water.nuclide}")

Hydrogen by symbol: ZAID=1000, fraction=2.0
Oxygen by ZAID: symbol=O, fraction=1.0

Has 'Fe'? False
Has 'H'? True


### 2.2 Density with Units

In [7]:
# density property returns just the float
print(f"Density (float): {water.density}")

# density_with_units property returns a formatted string
print(f"Density (with units): {water.density_with_units}")

# Convert density to different units
print(f"Density in kg/m3: {water.density_in('kg/m3')}")

Density (float): 1.0
Density (with units): 1 g/cc
Density in kg/m3: 1000.0


### 2.3 Fraction Type

In [8]:
print(f"Water fraction type: {water.fraction_type}")
print(f"Water is_atomic: {water.is_atomic}")
print(f"Water is_weight: {water.is_weight}")

print(f"\nSteel fraction type: {steel.fraction_type}")
print(f"Steel is_atomic: {steel.is_atomic}")
print(f"Steel is_weight: {steel.is_weight}")

Water fraction type: ao
Water is_atomic: True
Water is_weight: False

Steel fraction type: wo
Steel is_atomic: False
Steel is_weight: True


## 3. Modifying Materials

### 3.1 Setter Methods

In [9]:
# Create a test material
test_mat = Material(id=100)

# set_name()
test_mat.set_name('Test Material')
print(f"Name: {test_mat.name}")

# set_id()
test_mat.set_id(200)
print(f"ID: {test_mat.id}")

# set_library()
test_mat.set_library('nlib', '80c')
print(f"Libraries: {test_mat.libs}")

# remove_library()
test_mat.remove_library('nlib')
print(f"Libraries after removal: {test_mat.libs}")

Name: Test Material
ID: 200
Libraries: {'nlib': '80c'}
Libraries after removal: {}


### 3.2 Nuclide Modification Methods

In [10]:
# Create a material
mod_mat = Material(id=10)
mod_mat.add_nuclide('H', 1.0, 'ao')
mod_mat.add_nuclide('O', 1.0, 'ao')
mod_mat.add_nuclide('Fe', 1.0, 'ao')

print("Initial nuclides:")
for sym, nuc in mod_mat.nuclide.items():
    print(f"  {sym}: {nuc.fraction}")

# update_nuclide_fraction()
mod_mat.update_nuclide_fraction('H', 2.0)
print(f"\nAfter update_nuclide_fraction('H', 2.0):")
print(f"  H: {mod_mat.nuclide['H'].fraction}")

# remove_nuclide()
mod_mat.remove_nuclide('Fe')
print(f"\nAfter remove_nuclide('Fe'):")
for sym, nuc in mod_mat.nuclide.items():
    print(f"  {sym}: {nuc.fraction}")

# clear_nuclides()
mod_mat.clear_nuclides()
print(f"\nAfter clear_nuclides():")
print(f"  Number of nuclides: {len(mod_mat.nuclide)}")

Initial nuclides:
  H: 1.0
  O: 1.0
  Fe: 1.0

After update_nuclide_fraction('H', 2.0):
  H: 2.0

After remove_nuclide('Fe'):
  H: 2.0
  O: 1.0

After clear_nuclides():
  Number of nuclides: 0


### 3.3 copy() - Create Modified Versions

In [11]:
# Create a copy of water with a new ID
heavy_water = water.copy(new_id=11)
heavy_water.set_name('Heavy Water')

# Modify the copy (replace H with D)
heavy_water.remove_nuclide('H')
heavy_water.add_nuclide(1002, 2.0, 'ao')  # Deuterium
heavy_water.set_density(1.1, 'g/cc')

print(f"Original water ID: {water.id}, nuclides: {list(water.nuclide.keys())}")
print(f"Heavy water ID: {heavy_water.id}, nuclides: {list(heavy_water.nuclide.keys())}")

Original water ID: 1, nuclides: ['H', 'O']
Heavy water ID: 11, nuclides: ['O', 'H2']


## 4. Fraction Conversions

### 4.1 to_atomic_fraction() / to_weight_fraction()

In [12]:
# Create water with atomic fractions
water_test = Material(id=20)
water_test.add_nuclide('H', 2.0, 'ao')
water_test.add_nuclide('O', 1.0, 'ao')

print("Original (atomic fractions):")
print(f"  Type: {water_test.fraction_type}")
for sym, nuc in water_test.nuclide.items():
    print(f"  {sym}: {nuc.fraction:.6f}")

# Convert to weight fractions
water_test.to_weight_fraction()
print("\nAfter to_weight_fraction():")
print(f"  Type: {water_test.fraction_type}")
for sym, nuc in water_test.nuclide.items():
    print(f"  {sym}: {nuc.fraction:.6f}")

# Convert back to atomic fractions
water_test.to_atomic_fraction()
print("\nAfter to_atomic_fraction():")
print(f"  Type: {water_test.fraction_type}")
for sym, nuc in water_test.nuclide.items():
    print(f"  {sym}: {nuc.fraction:.6f}")

Original (atomic fractions):
  Type: ao
  H: 2.000000
  O: 1.000000

After to_weight_fraction():
  Type: wo
  H: 0.111115
  O: 0.888885

After to_atomic_fraction():
  Type: ao
  H: 0.666667
  O: 0.333333


### 4.2 normalize() - Scale Fractions to Sum=1

In [13]:
# Create a material with arbitrary fractions
h2o = Material(id=21)
h2o.add_nuclide('H', 2.0, 'ao')  # Not normalized (sum = 3.0)
h2o.add_nuclide('O', 1.0, 'ao')

print("Before normalize():")
total = sum(nuc.fraction for nuc in h2o.nuclide.values())
print(f"  Total: {total}")
for sym, nuc in h2o.nuclide.items():
    print(f"  {sym}: {nuc.fraction:.6f}")

# Normalize to atomic fractions (sum = 1.0)
h2o.normalize('ao')

print("\nAfter normalize('ao'):")
total = sum(nuc.fraction for nuc in h2o.nuclide.values())
print(f"  Total: {total}")
for sym, nuc in h2o.nuclide.items():
    print(f"  {sym}: {nuc.fraction:.6f}")

Before normalize():
  Total: 3.0
  H: 2.000000
  O: 1.000000

After normalize('ao'):
  Total: 1.0
  H: 0.666667
  O: 0.333333


### 4.3 to_integer_fractions() - Convert to Integer Ratios (NEW)

This method converts normalized fractions back to integer ratios. For example, H2O with fractions (H: 0.6667, O: 0.3333) becomes (H: 2, O: 1).

In [14]:
# Create a normalized material
mat_norm = Material(id=22)
mat_norm.add_nuclide('H', 2.0/3.0, 'ao')  # Normalized H2O
mat_norm.add_nuclide('O', 1.0/3.0, 'ao')

print("Normalized fractions:")
for sym, nuc in mat_norm.nuclide.items():
    print(f"  {sym}: {nuc.fraction:.6f}")

# Convert to integer fractions
mat_norm.to_integer_fractions()

print("\nAfter to_integer_fractions():")
for sym, nuc in mat_norm.nuclide.items():
    print(f"  {sym}: {nuc.fraction:.1f}")

Normalized fractions:
  H: 0.666667
  O: 0.333333

After to_integer_fractions():
  H: 2.0
  O: 1.0


In [15]:
# More complex example: sulfuric acid H2SO4
h2so4 = Material(id=23)
h2so4.add_nuclide('H', 2.0, 'ao')
h2so4.add_nuclide('S', 1.0, 'ao')
h2so4.add_nuclide('O', 4.0, 'ao')
h2so4.normalize('ao')

print("H2SO4 normalized:")
for sym, nuc in sorted(h2so4.nuclide.items()):
    print(f"  {sym}: {nuc.fraction:.6f}")

h2so4.to_integer_fractions()
print("\nH2SO4 integer fractions:")
for sym, nuc in sorted(h2so4.nuclide.items()):
    print(f"  {sym}: {nuc.fraction:.0f}")

H2SO4 normalized:
  H: 0.285714
  O: 0.571429
  S: 0.142857

H2SO4 integer fractions:
  H: 2
  O: 4
  S: 1


## 5. Natural Element Expansion

### 5.1 expand_natural_elements() - All Elements

In [16]:
# Create a material with natural elements
nat_mat = Material(id=30)
nat_mat.add_nuclide('H', 2.0, 'ao')
nat_mat.add_nuclide('O', 1.0, 'ao')

print("Before expansion:")
for sym, nuc in nat_mat.nuclide.items():
    print(f"  {sym} (ZAID {nuc.zaid}): {nuc.fraction:.6f}")

# Expand all natural elements
nat_mat.expand_natural_elements()

print("\nAfter expand_natural_elements():")
for sym, nuc in sorted(nat_mat.nuclide.items()):
    print(f"  {sym} (ZAID {nuc.zaid}): {nuc.fraction:.6e}")

Before expansion:
  H (ZAID 1000): 2.000000
  O (ZAID 8000): 1.000000

After expand_natural_elements():
  H (ZAID 1000): 2.000000e+00
  O16 (ZAID 8016): 9.975700e-01
  O17 (ZAID 8017): 3.800000e-04
  O18 (ZAID 8018): 2.050000e-03


### 5.2 expand_natural_elements() - Selective Expansion

In [17]:
# Create steel with natural elements
steel2 = Material(id=31)
steel2.add_element('Fe', 0.70, 'ao')
steel2.add_element('Cr', 0.18, 'ao')
steel2.add_element('Ni', 0.10, 'ao')
steel2.add_element('C', 0.02, 'ao')

# Expand only Fe
steel2.expand_natural_elements('Fe')
print("After expanding only Fe:")
for sym, nuc in sorted(steel2.nuclide.items(), key=lambda x: x[1].zaid):
    print(f"  {sym}: {nuc.fraction:.6e}")

After expanding only Fe:
  C: 2.000000e-02
  Cr: 1.800000e-01
  Fe54: 4.091500e-02
  Fe56: 6.422780e-01
  Fe57: 1.483300e-02
  Fe58: 1.974000e-03
  Ni: 1.000000e-01


In [18]:
# Expand specific elements with a list
steel3 = Material(id=32)
steel3.add_element('Fe', 0.70, 'ao')
steel3.add_element('Cr', 0.18, 'ao')
steel3.add_element('Ni', 0.10, 'ao')
steel3.add_element('C', 0.02, 'ao')

# Expand Fe and C only
steel3.expand_natural_elements(['Fe', 'C'])
print("After expanding Fe and C:")
for sym, nuc in sorted(steel3.nuclide.items(), key=lambda x: x[1].zaid):
    print(f"  {sym}: {nuc.fraction:.6e}")

After expanding Fe and C:
  C12: 1.978600e-02
  C13: 2.140000e-04
  Cr: 1.800000e-01
  Fe54: 4.091500e-02
  Fe56: 6.422780e-01
  Fe57: 1.483300e-02
  Fe58: 1.974000e-03
  Ni: 1.000000e-01


## 6. Code Output Formats

### 6.1 to_mcnp() - MCNP Material Cards

In [19]:
# Create a complete material for MCNP output
mcnp_mat = Material(id=100)
mcnp_mat.set_name('Test Material')
mcnp_mat.add_nuclide('U235', 0.05, 'ao', '80c')  # With library
mcnp_mat.add_nuclide('U238', 0.95, 'ao', '80c')
mcnp_mat.set_density(19.1, 'g/cc')

print("MCNP output:")
print(mcnp_mat.to_mcnp())

MCNP output:
c KIKA_MAT_NAME: Test Material
m100  $ KIKA_DENSITY: 1.910000e+01 g/cc, 4.835942e-02 atoms/b-cm
     92235.80c      5.000000e-02   $ U235  
     92238.80c      9.500000e-01   $ U238  


### 6.2 to_serpent() - Serpent Material Definition (NEW)

In [20]:
# Same material in Serpent format
print("Serpent output:")
print(mcnp_mat.to_serpent())

Serpent output:
mat Test_Material -1.910000e+01
    92235.80c      5.000000e-02
    92238.80c      9.500000e-01


In [21]:
# Serpent output with temperature
hot_mat = Material(id=101)
hot_mat.set_name('Hot Water')
hot_mat.add_nuclide('H', 2.0, 'ao', '80c')
hot_mat.add_nuclide('O', 1.0, 'ao', '80c')
hot_mat.normalize()
hot_mat.set_density(0.95, 'g/cc')
hot_mat.set_temperature(600.0)  # 600 K

print("Serpent output with temperature:")
print(hot_mat.to_serpent())

Serpent output with temperature:
mat Hot_Water -9.500000e-01 tmp 600.0
    1000.80c       6.666667e-01
    8000.80c       3.333333e-01


### 6.3 Weight Fraction Output

In [22]:
# Weight fraction material
wo_mat = Material(id=102)
wo_mat.add_nuclide('Fe', 0.70, 'wo')
wo_mat.add_nuclide('Cr', 0.18, 'wo')
wo_mat.add_nuclide('Ni', 0.12, 'wo')
wo_mat.set_density(8.0, 'g/cc')

print("MCNP output (weight fractions):")
print(wo_mat.to_mcnp())
print("\nSerpent output (weight fractions):")
print(wo_mat.to_serpent())

MCNP output (weight fractions):
m102  $ KIKA_DENSITY: 8.000000e+00 g/cc, 8.691627e-02 atoms/b-cm
     24000         -1.800000e-01   $ Cr    
     26000         -7.000000e-01   $ Fe    
     28000         -1.200000e-01   $ Ni    

Serpent output (weight fractions):
mat mat102 -8.000000e+00
    24000         -1.800000e-01
    26000         -7.000000e-01
    28000         -1.200000e-01


## 7. Material Collections

### 7.1 Creating a Collection

In [23]:
# Create materials
mat1 = Material(id=1)
mat1.set_name('Water')
mat1.add_nuclide('H', 2.0, 'ao')
mat1.add_nuclide('O', 1.0, 'ao')
mat1.normalize()
mat1.set_density(1.0, 'g/cc')

mat2 = Material(id=2)
mat2.set_name('Uranium')
mat2.add_nuclide(92235, 0.05, 'ao')
mat2.add_nuclide(92238, 0.95, 'ao')
mat2.set_density(19.1, 'g/cc')

# Create collection and add materials
collection = MaterialCollection()
collection.add_material(mat1)
collection.add_material(mat2)

print(collection)

                              Materials Collection                              

Total Materials: 2

--------------------------------------------------------------------------------
     ID      |  Nuclides  |    Type    |               Libraries               
--------------------------------------------------------------------------------
     1       |     2      |   Atomic   |                default                
     2       |     2      |   Atomic   |                default                
--------------------------------------------------------------------------------

Available methods:
  .add_material(material)  - Add a material to the collection
  .to_mcnp()               - Export all materials as MCNP cards
  .to_serpent()            - Export all materials as Serpent cards
  .by_id[material_id]      - Access a specific material by ID

Examples of accessing data:
  collection.by_id[1]      - Access material with ID 1
  for mat in collection:   - Iterate over all materials


### 7.2 Accessing Materials in Collection

In [24]:
# Access by ID
water = collection.by_id[1]
print(f"Material 1: {water.name}")

# Iterate over all materials
print("\nAll materials:")
for mat in collection:
    print(f"  ID {mat.id}: {mat.name}")

Material 1: Water

All materials:
  ID 1: Water
  ID 2: Uranium


### 7.3 Collection Export

In [25]:
# Export all materials as MCNP cards
print("=== MCNP Output ===")
print(collection.to_mcnp())

=== MCNP Output ===
c KIKA_MAT_NAME: Water
m1  $ KIKA_DENSITY: 1.000000e+00 g/cc, 1.003723e-01 atoms/b-cm
     1000           6.666667e-01   $ H     
     8000           3.333333e-01   $ O     
c KIKA_MAT_NAME: Uranium
m2  $ KIKA_DENSITY: 1.910000e+01 g/cc, 4.835942e-02 atoms/b-cm
     92235          5.000000e-02   $ U235  
     92238          9.500000e-01   $ U238  


In [26]:
# Export all materials as Serpent cards
print("=== Serpent Output ===")
print(collection.to_serpent())

=== Serpent Output ===
mat Water -1.000000e+00
    1000           6.666667e-01
    8000           3.333333e-01

mat Uranium -1.910000e+01
    92235          5.000000e-02
    92238          9.500000e-01


### 7.4 Export to File

In [27]:
# Export to file (MCNP format)
# collection.to_file('materials.txt')  # All materials
# collection.to_file('selected.txt', material_ids=[1])  # Only specific materials

print("File export methods available:")
print("  collection.to_file('materials.txt')")
print("  collection.to_file('selected.txt', material_ids=[1, 2])")
print("  collection.write_to_mcnp('input.i', 'output.i')")

File export methods available:
  collection.to_file('materials.txt')
  collection.to_file('selected.txt', material_ids=[1, 2])
  collection.write_to_mcnp('input.i', 'output.i')


## 8. Summary of New Features

The materials module has been restructured and enhanced with the following new features:

### Module Structure
- Moved from `kika.mcnp.material` to `kika.materials` (with backwards compatibility)
- Supports multi-code output (MCNP, Serpent)

### New Properties
- `density_with_units` - Returns density as a formatted string with units

### New Methods
- `to_serpent()` - Export material as Serpent format
- `to_integer_fractions()` - Convert normalized fractions to integer ratios
- `set_name()` - Set or change material name
- `set_id()` - Change material ID
- `set_library()` - Set material-level library defaults
- `remove_library()` - Remove material-level library setting
- `remove_nuclide()` - Remove a nuclide from the material
- `update_nuclide_fraction()` - Update fraction of existing nuclide
- `clear_nuclides()` - Remove all nuclides

### Collection Methods
- `to_serpent()` - Export all materials as Serpent format